packages/@aws-cdk/aws-ec2/lib/client-vpn-endpoint.ts

```diff
@@ -0,0 +1,385 @@
+  readonly logGroup?: logs.ILogGroup;
+
+  /**
+   * A CloudWatch Logs log stream for connection logging
+   *
+   * @default - a new stream is created
+   */
+  readonly logStream?: logs.ILogStream;
+
+  /**
+   * The AWS Lambda function used for connection authorization
+   *
+   * @default - no connection handler
+   */
+  readonly clientConnectionHandler?: IFunction; // Redefined IFunction to avoid circular dependency
+
+  /**
+   * A brief description of the Client VPN endpoint.
+   *
+   * @default - no description
+   */
+  readonly description?: string;
+
+  /**
+   * The security groups to apply to the target network.
+   *
+   * @default - a new security group is created
+   */
+  readonly securityGroups?: ISecurityGroup[];
+
+  /**
+   * Specify whether to enable the self-service portal for the Client VPN endpoint.
+   *
+   * @default true
+   */
+  readonly selfServicePortal?: boolean;
+
+  /**
+   * The server certificate
+   */
+  readonly serverCertificate: ICertificate;
+
+  /**
+   * Indicates whether split-tunnel is enabled on the AWS Client VPN endpoint.
+   *
+   * @see https://docs.aws.amazon.com/vpn/latest/clientvpn-admin/split-tunnel-vpn.html
+   *
+   * @default false
+   */
+  readonly splitTunnel?: boolean;
+
+  /**
+   * The transport protocol to be used by the VPN session.
+   *
+   * @default TransportProtocol.UDP
+   */
+  readonly transportProtocol?: TransportProtocol;
+
+  /**
+   * The port number to assign to the Client VPN endpoint for TCP and UDP
+   * traffic.
+   *
+   * @default VpnPort.HTTPS
+   */
+  readonly port?: VpnPort;
+
+  /**
+   * Information about the DNS servers to be used for DNS resolution.
+   *
+   * A Client VPN endpoint can have up to two DNS servers.
+   *
+   * @default - use the DNS address configured on the device
+   */
+  readonly dnsServers?: string[];
+
+  /**
+   * Subnets to associate to the client VPN endpoint.
+   *
+   * @default - the VPC default strategy
+   */
+  readonly vpcSubnets?: SubnetSelection;
+
+  /**
+   * Whether to authorize all users to the VPC CIDR
+   *
+   * This automatically creates an authorization rule. Set this to `false` and
+   * use `addAuthorizationRule()` to create your own rules instead.
+   *
+   * @default true
+   */
+  readonly authorizeAllUsersToVpcCidr?: boolean;
+}
+
+/**
+ * User-based authentication for a client VPN endpoint
+ */
+export abstract class ClientVpnUserBasedAuthentication {
+  /**
+   * Active Directory authentication
+   */
+  public static activeDirectory(directoryId: string): ClientVpnUserBasedAuthentication {
+    return { directoryId };
+  }
+
+  /**
+   * Federated authentication
+   */
+  public static federated(samlProviderArn: string, selfServiceSamlProviderArn?: string): ClientVpnUserBasedAuthentication {
+    return {
+      samlProviderArn,
+      selfServiceSamlProviderArn,
+    };
+  }
+
+  /** The ID of the Active Directory to be used for authentication */
+  public abstract readonly directoryId?: string;
```

Comment:
Feels more correct to abstract this away one more level and return the entire object, with "whatever" fields need to be in there fore the given auth type.

Reply:
You mean something like this?

```ts
public abstract readonly directory?: { directoryId: string };
public abstract readonly samlProvider?: { samlProviderArn: string, selfServiceSamlProviderArn?: string };
```

Reply:
Actually I don't get it, isn't the `ClientVpnUserBasedAuthentication` the entire object with the fields set according to the chosen auth?

Reply:
It is. I'm talking about the protocol between `ClientVpnUserBasedAuthentication` and `ClientVpnEndpoint`.

I was thinking the protocol didn't need to be more complicated than this:

```ts
export abstract class ClientVpnUserBasedAuthentication {
  public abstract render(): any;
}
```

Then `ClientVpnEndpoint` doesn't need to care about the individual fields in that output structure anymore, but can just trust that whatever `ClientVpnUserBasedAuthentication` returns is correct.

Reply:
OK, this means 

```ts
export class ClientVpnActiveDirectoryAuthentication extends ClientVpnUserBasedAuthentication {
  constructor(private readonly directoryId: string) {
    super();
  }

  render() {
    return {
      type: 'directory-service-authentication',
      activeDirectory: {
        directoryId: this.directoryId,
      },
    };
  }
}
```

and we drop the static methods on `ClientVpnUserBasedAuthentication`?

Reply:
The static methods still make sense as factory functions. You can also skip exporting the class that way.